### PR TITLE
[2] feat: Sobolev training

### PR DESCRIPTION
**Pull request by WojtAcht**
TODO:
- [x] ~~identify original repo (XD)~~,
- [x] ~~test on a new larger dataset~~,
- [x] compare it against MLP,
- [x] If it works, integrate it with the rest (MLP, TensorBoard, Lightning, evaluators, etc.). I attempted to implement it using Lightning, but my implementation failed to converge (you can check the implementation in older commits on this branch).
- [x] how can we run this on BBOB? Is it possible to efficiently extract gradients?
- [x] check this implementation: https://github.com/AberHu/Knowledge-Distillation-Zoo/blob/master/kd_losses/sobolev.py it's super clean, maybe we can use it.

Paper: https://arxiv.org/abs/1706.04859

### CODE REVIEW

File: nn_surrogate_benchmark/bbob_sampler.py [hunk 1]
```diff
@@ -27,16 +28,17 @@ def generate_lhs_samples(
 
 def main():
     BOUNDS = [-5.0, 5.0]
-    N_SAMPLES = 10_000
+    N_SAMPLES = 30_000
+    GET_GRADIENTS = True
     DIMENSIONS = 2
     INSTANCE_ID = 1
-    FUNCTION_ID = 1
+    FUNCTION_ID = 24
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/bbob_sampler.py:35]: We should refactor these variables into a configuration class or file, though it’s fine for now.

File: nn_surrogate_benchmark/bbob_sampler.py [hunk 2]
```diff
@@ -27,16 +28,17 @@ def generate_lhs_samples(
...
     DATA_DIR = "data"
 
     suite = Suite(
         "bbob",
-        f"instances: {FUNCTION_ID}",
-        f"dimensions: {DIMENSIONS} instance_indices: {INSTANCE_ID}",
+        "",
+        f"function_indices: {FUNCTION_ID}, dimensions: {DIMENSIONS} instance_indices: {INSTANCE_ID}",
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/bbob_sampler.py:41]: thanks, I forgot to fix this! 👏🏿

File: nn_surrogate_benchmark/bbob_sampler.py [hunk 3]
```diff
@@ -53,17 +55,29 @@ def main():
...
-            rows.append(
-                {
-                    **{f"x{i+1}": float(val) for i, val in enumerate(x)},
-                    "y": float(f_value),
-                }
-            )
+        if GET_GRADIENTS:
+            for x in tqdm(combinations):
+                f_value = function(x)
+                grad = approx_fprime(x, function, epsilon=1e-5)
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/bbob_sampler.py:61]: It would be nice to implement BBOB functions using torch e.g.
```
def rosenbrock_function(x: torch.Tensor, a: float = 1.0, b: float = 100.0) -> torch.Tensor:
    """
    2D Rosenbrock function.
    Global minimum at (1,1) with f(1,1)=0.
    """
    return (a - x[0])**2 + b * (x[1] - x[0]**2)**2
```
you can find already existing repo here: https://github.com/PatrikValkovic/BBOBtorch

File: nn_surrogate_benchmark/ela_comparator.py [hunk 4]
```diff
@@ -49,9 +49,8 @@ def _extract_from_dataloader(
         y_list: list[np.ndarray] = []
         total_points = 0
 
-        for X_batch, y_batch in dataloader:
-            batch_X = self._prepare_data(X_batch)[0]
-            batch_y = self._prepare_data(y_batch)[0]
+        for batch in dataloader:
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/ela_comparator.py:52]: thanks, good catch!

File: nn_surrogate_benchmark/model_evaluator.py [hunk 5]
```diff
@@ -97,6 +99,7 @@ def evaluate_multiple_sets(
         train_loader: DataLoader | None = None,
         val_loader: DataLoader | None = None,
         test_loader: DataLoader | None = None,
+        grads = False,
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/model_evaluator.py:102]:
```
grads: bool = False
```

File: nn_surrogate_benchmark/surrogate_sobol.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+from torch.utils.data import TensorDataset, DataLoader
+from sklearn.preprocessing import (
+	StandardScaler,
+	MinMaxScaler,
+	RobustScaler,
+	MaxAbsScaler,
+)
+
+class Sobolev(pl.LightningModule):
+	def __init__(
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/surrogate_sobol.py:17]: This formatting differs from my default; I need to properly set up pre-commit.

File: nn_surrogate_benchmark/surrogate_sobol.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+import numpy as np
+from torch.utils.data import TensorDataset, DataLoader
+from sklearn.preprocessing import (
+	StandardScaler,
+	MinMaxScaler,
+	RobustScaler,
+	MaxAbsScaler,
+)
+
+class Sobolev(pl.LightningModule):
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/surrogate_sobol.py:16]: maybe it could inherit from MLP? this way it'll be easier to maintain

File: nn_surrogate_benchmark/surrogate_sobol.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+		super().__init__()
+		self.save_hyperparameters()
+
+		layers = []
+		for hidden_dim in hidden_dims:
+			layers.append(nn.Linear(input_dim, hidden_dim))
+			if activation == "tanh":
+				layers.append(nn.Tanh())
+			elif activation == "relu":
+				layers.append(nn.ReLU(inplace=False))  # Ensure inplace=False
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/surrogate_sobol.py:34]: why `inplace=False`? we don't have this for MLP

File: nn_surrogate_benchmark/surrogate_sobol.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+	def test_step(self, batch, batch_idx):
+		x, y = batch
+		preds = self(x)
+		loss = self.criterion(preds, y)
+		self.log("test_loss", loss, on_epoch=True, prog_bar=True)
+		return loss
+
+	def configure_optimizers(self):
+		optimizer = Adam(self.parameters(), lr=self.hparams.lr, weight_decay=1e-5)
+		scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=300, gamma=0.5)
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/surrogate_sobol.py:103]: for MLP we have:
```
    def configure_optimizers(self):
        optimizer = AdamW(self.parameters(), lr=self.hparams.lr)
        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
            optimizer, mode="min", factor=0.1, patience=5, verbose=True
        )

        return {
            "optimizer": optimizer,
            "lr_scheduler": {"scheduler": scheduler, "monitor": "val_loss"},
        }
```
why do use StepLR here? Does it work better?

File: nn_surrogate_benchmark/model_evaluator.py [hunk 10]
```diff
@@ -67,7 +67,9 @@ def evaluate_dataset(
         dataset_name: str,
     ) -> pd.DataFrame:
         X, y_true_scaled = self._extract_from_dataloader(dataloader)
+        y_true_scaled = y_true_scaled.flatten()
```
Comment (WojtAcht) [inline nn_surrogate_benchmark/model_evaluator.py:70]: why?

File: nn_surrogate_benchmark/bbob_sampler.py (hunk 3 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/bbob_sampler.py:61]: Nice idea! Although we are moving the BBOB sampler logic to the second repo, so the changes will be applied there

File: nn_surrogate_benchmark/model_evaluator.py (hunk 5 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/model_evaluator.py:102]: I removed the variable completely as it was not used

File: nn_surrogate_benchmark/surrogate_sobol.py (hunk 6 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/surrogate_sobol.py:17]: Ok, for the moment I tried unifying the formatting

File: nn_surrogate_benchmark/surrogate_sobol.py (hunk 9 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/surrogate_sobol.py:103]: Yes, I arrived at StepLR through trial and error

File: nn_surrogate_benchmark/model_evaluator.py (hunk 10 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/model_evaluator.py:70]: Before that I was getting dimension mismatch errors. It seemed related to how the loader structured the batch data

File: nn_surrogate_benchmark/surrogate_sobol.py (hunk 7 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/surrogate_sobol.py:16]: I'm not well versed enough to have an opinion on this choice. For now pytorch lightning seems fine for me

File: nn_surrogate_benchmark/surrogate_sobol.py (hunk 8 above)
Comment (Hubert-Guzowski) [reply nn_surrogate_benchmark/surrogate_sobol.py:34]: Yeah, it's unneeded - inplace=False is the default. Removed it
